test/lietensor/test_lietensor.py

```diff
@@ -0,0 +1,297 @@
+import time
+import torch
+import random
+import warnings
+import torch.linalg
+import pypose as pp
+from torch import nn
+import pypose.optim.solver as ppos
+import pypose.optim.kernel as ppok
+import pypose.optim.corrector as ppoc
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+
+def test_lietensor():
+
+    a = torch.randn(3,3)*0.1
+    x = pp.so3(a)
+    y = pp.randn_so3(3, sigma=0.1, dtype=torch.float64, requires_grad=True, device=device)
+    a = pp.randn_se3(3, sigma=0.1, requires_grad=True)
+    b = pp.SE3_type.randn(3, sigma=0.1, requires_grad=True)
+
+    assert y.is_leaf and x.is_leaf and a.is_leaf and b.is_leaf
+
+    (pp.Log(y.Exp())**2).sin().sum().backward()
+
+    print(y)
+    print(y.lshape)
+    print(y.grad)
+
+    def count_parameters(model):
+        return sum(p.numel() for p in model.parameters() if p.requires_grad)
+
+    class SO3Layer(nn.Module):
+        def __init__(self, n):
+            super().__init__()
+            self.weight = pp.Parameter(pp.randn_so3(n))
+
+        def forward(self, x):
+            return self.weight.Exp() * x
+
+    n = 4
+    epoch = 3
+    net = SO3Layer(n).to(device)
+    print(net.weight)
+    net.weight.data.add_(0.1)
+    print(net.weight)
+    optimizer = torch.optim.SGD(net.parameters(), lr = 0.1, momentum=0.9)
+    scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[10,15], gamma=0.1)
+    print(net.weight)
+    for i in range(epoch):
+        optimizer.zero_grad()
+        inputs = pp.randn_SO3(n, device=device)
+        outputs = net(inputs)
+        loss = outputs.abs().sum()
+        loss.backward()
+        optimizer.step()
+        scheduler.step()
+        print(loss)
+    print(net.weight)
+    print("Parameter:", count_parameters(net))
+
+    SO3I = pp.identity_SO3(1, 3, device=device, dtype=torch.float64)
```

Comment:
It may be more readable and scalable to use [@pytest.mark.parametrize](https://docs.pytest.org/en/7.1.x/how-to/parametrize.html) to generate a cartesian product of test inputs (`identity_<ltype>` x shape in this case).

Reply:
This is to test whether the grammar works. not works for specific data.